src/preprocessing/particle_packing/system.jl

```diff
@@ -0,0 +1,283 @@
+    return sum(each_moving_particle(system), init=zero(eltype(system))) do particle
+        velocity = extract_svector(initial_condition.velocity, system, particle)
+        return 0.5 * initial_condition.mass[particle] * dot(velocity, velocity)
+    end
+end
+
+@inline source_terms(system::ParticlePackingSystem) = nothing
+@inline add_acceleration!(dv, particle, system::ParticlePackingSystem) = dv
+
+# Number of particles in the system
+@inline nparticles(system::ParticlePackingSystem) = length(system.initial_condition.mass)
+
+@inline function hydrodynamic_mass(system::ParticlePackingSystem, particle)
+    return system.initial_condition.mass[particle]
+end
+
+update_particle_packing(system, v_ode, u_ode, semi, integrator) = system
+
+function update_particle_packing(system::ParticlePackingSystem, v_ode, u_ode,
+                                 semi, integrator)
+    u = wrap_u(u_ode, system, semi)
+
+    update_position!(u, system)
+end
+
+function update_position!(u, system::ParticlePackingSystem)
+    func_name = "constrain outside particles onto surface"
+    @trixi_timeit timer() func_name constrain_particles_onto_surface!(u, system)
+
+    return u
+end
+
+function update_final!(system::ParticlePackingSystem, v, u, v_ode, u_ode, semi, t;
+                       update_from_callback=false)
+    if !update_from_callback && !(system.update_callback_used[])
+        throw(ArgumentError("`UpdateCallback` is required when using `ParticlePackingSystem`"))
+    end
+
+    return system
+end
+
+function constrain_particles_onto_surface!(u, system::ParticlePackingSystem)
+    (; neighborhood_search, signed_distance_field) = system
+    (; positions, distances, normals) = signed_distance_field
+
+    search_radius2 = compact_support(system, system)^2
+
+    @threaded system for particle in eachparticle(system)
+        particle_position = current_coords(u, system, particle)
+
+        volume = zero(eltype(system))
+        distance_signed = zero(eltype(system))
+        normal_vector = fill(volume, SVector{ndims(system), eltype(system)})
+
+        # Interpolate signed distances and normals
+        for neighbor in PointNeighbors.eachneighbor(particle_position, neighborhood_search)
+            pos_diff = positions[neighbor] - particle_position
+            distance2 = dot(pos_diff, pos_diff)
+            distance2 > search_radius2 && continue
+
+            distance = sqrt(distance2)
+            kernel_weight = smoothing_kernel(system, distance)
+
+            distance_signed += distances[neighbor] * kernel_weight
+
+            normal_vector += normals[neighbor] * kernel_weight
+
+            volume += kernel_weight
+        end
+
+        if volume > eps()
+            distance_signed /= volume
+            normal_vector /= volume
+
+            # Store signed distance for visualization
+            system.signed_distances[particle] = distance_signed
+
+            constrain_particles!(u, system, particle, distance_signed, normal_vector)
```

Comment:
```suggestion
            constrain_particle!(u, system, particle, distance_signed, normal_vector)
```
This is only constraining `particle`.